### PR TITLE
Unify meta tensor and fake tensor converter conversion

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87943
* #87911
* #87903

Meta tensor does a lot of work to make sure tensors "look" similar
to the original parts; e.g., if the original was a non-leaf, meta
converter ensures the meta tensor is a non-leaf too.  Fake tensor
destroyed some of these properties when it wraps it in a FakeTensor.

This patch pushes the FakeTensor constructor into the meta converter
itself, so that we first create a fake tensor, and then we do various
convertibility bits to it to make it look right.

The two tricky bits:

- We need to have no_dispatch enabled when we allocate the initial meta
  tensor, or fake tensor gets mad at us for making a meta fake tensor.
  This necessitates the double-callback structure of the callback
  arguments: the meta construction happens *inside* the function so
  it is covered by no_dispatch

- I can't store tensors for the storages anymore, as that will result
  in a leak.  But we have untyped storage now, so I just store untyped
  storages instead.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @jansel @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx

### CODE REVIEW

File: torch/_subclasses/meta_utils.py [hunk 1]
```diff
@@ -280,10 +295,13 @@ def __call__(self, t, shape_env=None):
...
+                    return NotImplemented
                 return t
             else:
                 self.hit += 1
-                r = self.meta_tensor(t, shape_env=shape_env)
+                r = self.meta_tensor(t, shape_env=shape_env, callback=callback)
+                # TODO: this is suspicious, now that we have callback argument
                 if type(t) is torch.nn.Parameter:
                     r = torch.nn.Parameter(r, requires_grad=r.requires_grad)
                 return r
```
Comment (eellison) [inline torch/_subclasses/meta_utils.py:307]: Should we move this logic https://github.com/pytorch/pytorch/blob/48e1981d0286c75afe043750c3f3b4f6b841405c/torch/_subclasses/fake_tensor.py#L235-L239 here ?

File: torch/_subclasses/meta_utils.py [hunk 2]
```diff
@@ -127,18 +127,24 @@ def del_ten():
...
         # NB: TypedStorage is freshly allocated and cannot be used as hash
         # key index.
 
         # Use a Weak Ref to s in order to not leak memory
         swr = StorageWeakRef(s)
         if swr not in self.storage_memo:
-            self.storage_memo[swr] = torch.empty(s.size(), dtype=s.dtype, device="meta")
+            self.storage_memo[swr] = (
+                callback(
+                    lambda: torch.empty(s.size(), dtype=torch.uint8, device="meta")
```
Comment (eellison) [inline torch/_subclasses/meta_utils.py:139]: `s.dtype ->torch.uint8` - don't we need `s.size() * s.element_size()` if we're going to use `uint8`

**Review by eellison — COMMENTED** [review]
Cool! just asking some questions and want to make sure I understand. 

> I can't store tensors for the storages anymore, as that will result
in a leak. But we have untyped storage now, so I just store untyped
storages instead.

Why is this ?

File: torch/_subclasses/meta_utils.py [hunk 3]
```diff
@@ -229,11 +242,13 @@ def is_c_of_r(complex_dtype, real_dtype):
                                 # sort of unsupported grad_fn here
                                 r = r.clone()
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:243]: I changed this: fake tensor mode *is* active here. So r will be fake

File: torch/_subclasses/meta_utils.py [hunk 4]
```diff
@@ -229,11 +242,13 @@ def is_c_of_r(complex_dtype, real_dtype):
...
-                    # redispatching on set_(Storage, ...) which will choke with
-                    # meta storage
-                    s = self.meta_storage(t.storage())
+                        r = callback(
+                            lambda: torch.empty((0,), dtype=t.dtype, device="meta")
+                        )
+                        assert safe_is_leaf(
+                            r
+                        ), "the callback you passed in doesn't detach"
+                    s = self.meta_storage(t.storage().untyped(), callback=callback)
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:251]: Without the callback, fake tensor mode is active and I attempt to create a meta fake tensor which is not allowed by default. So I must use the callback to instead make a non meta storage (which actually is meta cuz fake mode).

File: torch/_subclasses/meta_utils.py (hunk 1 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:307]: Oh yeah, good catch

File: torch/_subclasses/meta_utils.py (hunk 2 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:139]: No because it's untyped storage so already reporting size in bytes. Maybe could use nbytes() to ne clearer.

**Comment by ezyang** [comment]
Ran into a funny bug in autograd https://github.com/pytorch/pytorch/issues/87956 so I had to refactor meta converter a bit more to make it better.

**Comment by ezyang** [comment]
> Why is this ?

The storage map is a strong map and I didn't feel like changing this

File: torch/_subclasses/meta_utils.py (hunk 1 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:342]: I moved it into meta_tensor itself, instead of at top level. This way recursive grad field also works.

**Comment by ezyang** [comment]
So there was another subtle bug, where `no_dispatch` was on (due to fake tensor enabling it) when I attempt to clone in converter (to fake up autograd history). To fix this, I added an assert to converter that no dispatch is NOT disabled (invariant: you must always dispatch), and then refactored fake tensor to push the no dispatch / in kernel manager calls as close to the call sites as possible.

Once I did this, I also found that I had to setup in kernel manager for `set_` call. I break encapsulation and just do this by hand.

**Comment by ezyang** [comment]
If you desire a smaller patch, I could probably extract out the "make no dispatch extents smaller" into its own refactor patch, but I'd prefer not to.

**Comment by eellison** [comment]
Would it be worth considering adding an API to re-enable torch dispatch, or an api to disable it such that it could be re-enabled ? I think that might make the changes easier/less invasive. For instance, instead of using 'no_dispatch' directly, we could add the a context manager like in_kernel_invocation that adds the torch._C.DisableTorchDispatch object on the fake tensor mode itself such that a subsequent context manager could temporarily delete it

**Comment by ezyang** [comment]
I think that the newly refactored code is better. Previously, it seemed that we were indiscriminately spraying no_dispatch() inside the inner implementation of fake tensor. no_dispatch without in kernel mode is dangerous because any op you call can cause a segfault, as the fake tensor is still advertising that its a CPU tensor. It is better to keep the no_dispatch calls scoped as tightly as possible.

File: torch/_subclasses/fake_tensor.py [hunk 5]
```diff
@@ -655,6 +662,13 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
             else:
                 return args[0].fake_device
 
+        # Some attribute queries that can be serviced directly
+        # See Note [is_coalesced is dispatched]
+        if func in [torch.ops.aten.is_coalesced.default]:
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:667]: reason not to register this with `register_op_impl` ?

File: torch/_subclasses/meta_utils.py [hunk 6]
```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
...
+
+                    s = t.storage().untyped()
+                    swr = StorageWeakRef(s)
+                    if (
+                        swr not in self.storage_memo
+                        and r.stride() == strides
+                        and r.storage_offset() == storage_offset
+                    ):
+                        # You're normal and happy, install the fresh storage into the memo
+                        self.storage_memo[swr] = r.storage().untyped()
```
Comment (eellison) [inline torch/_subclasses/meta_utils.py:277]: seems a bit strange we have to directly invoke the memo here instead of as part of normal conversion

File: torch/_subclasses/meta_utils.py [hunk 7]
```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
...
+                if grad_not_none:
+                    r.grad = self.meta_tensor(t.grad, shape_env, callback)
                 torch._C._set_conj(r, t.is_conj())
                 torch._C._set_neg(r, t.is_neg())
             self.set_tensor_memo(t, r)
 
         return self.get_tensor_memo(t)
 
-    def __call__(self, t, shape_env=None):
+    def __call__(self, t, shape_env=None, *, strict=False, callback=lambda t: t()):
```
Comment (eellison) [inline torch/_subclasses/meta_utils.py:329]: the only other callee of this is `run_meta_crossref` - could remove the `strict=False` arg, always return NotImplemented, and remove this arg. not a big deal either way

File: torch/_subclasses/meta_utils.py [hunk 8]
```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
...
+                        if isinstance(r, FakeTensor):
+                            maybe_fake_mgr = in_kernel_invocation_manager(r.fake_mode)
+                        with maybe_fake_mgr, torch.no_grad():
+                            r.set_(r_s, storage_offset, sizes, strides)
 
+                with warnings.catch_warnings():
+                    warnings.filterwarnings("ignore", "The .grad attribute of a Tensor")
+                    grad_not_none = t.grad is not None
+                if grad_not_none:
+                    r.grad = self.meta_tensor(t.grad, shape_env, callback)
```
Comment (eellison) [inline torch/_subclasses/meta_utils.py:322]: What happens if the `grad` here isn't convertible to meta ? is that possible/something we need to handle ?

File: torch/_subclasses/meta_utils.py [hunk 9]
```diff
@@ -280,10 +361,13 @@ def __call__(self, t, shape_env=None):
...
                 self.miss += 1
+                if strict:
+                    return NotImplemented
                 return t
             else:
                 self.hit += 1
-                r = self.meta_tensor(t, shape_env=shape_env)
+                r = self.meta_tensor(t, shape_env=shape_env, callback=callback)
+                # TODO: this is suspicious, now that we have callback argument
                 if type(t) is torch.nn.Parameter:
```
Comment (eellison) [inline torch/_subclasses/meta_utils.py:371]: this is suspicious because the callback could call `torch.nn.Parameter` on the output instead of here ?

**Review by eellison — APPROVED** [review]
🚀

File: torch/_subclasses/meta_utils.py [hunk 10]
```diff
@@ -127,18 +129,31 @@ def del_ten():
...
+            )
         return self.storage_memo[swr]
 
     # This function assumes that it's possible to do the conversion
-    def meta_tensor(self, t, shape_env=None):
+    def meta_tensor(self, t, shape_env=None, callback=lambda t: t()):
+        # This indicates you set no_dispatch() before calling into this
+        # function.  This is an error: we may be creating fake tensors and
+        # will perform operations on them which need fake tensor mode to
+        # be active.  You will segfault if you are in a no_dispatch() block.
```
[the remark this answers is not in the document]
Comment (albanD) [reply torch/_subclasses/meta_utils.py:153]: Can we actually fix that? You can very easily segfault if you use no_dispatch. That is ~expected.

File: torch/_subclasses/meta_utils.py [hunk 11]
```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
...
-                            (0,), dtype=t.dtype, device="meta", requires_grad=True
+                        r = callback(
+                            lambda: torch.empty_strided(
+                                sizes, strides, dtype=t.dtype, device="meta"
+                            )
                         )
+                        assert safe_is_leaf(
+                            r
+                        ), "the callback you passed in doesn't detach"
+                        r.requires_grad = t.requires_grad
```
Comment (albanD) [inline torch/_subclasses/meta_utils.py:252]: Now that this is moved outside of the empty call, the logic here is:
```
if ..:
  do A
  do B
else:
  do A
```
We can clean that up.

File: torch/_subclasses/meta_utils.py [hunk 12]
```diff
@@ -214,38 +237,96 @@ def is_c_of_r(complex_dtype, real_dtype):
...
+                    if (
+                        swr not in self.storage_memo
+                        and r.stride() == strides
+                        and r.storage_offset() == storage_offset
+                    ):
+                        # You're normal and happy, install the fresh storage into the memo
+                        self.storage_memo[swr] = r.storage().untyped()
+                    else:
+                        # You're in crazy town; somehow you gave us a tensor
+                        # that wasn't a view, but had nonzero storage offset,
```
Comment (albanD) [inline torch/_subclasses/meta_utils.py:280]: nit: wasn't a view in what sense? "differentiable view" ?
If so, you can easily get here by detach-ing any weird Tensor you had.

**Review by albanD — APPROVED** [review]
Sounds good!

File: torch/_subclasses/meta_utils.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:153]: Yeah, it's not possible to fix, I think. The problem is that `no_dispatch` will strip the Python key, but if you don't add Meta to the include set, you'll just go straight to a CPU kernel with a fake tensor, and then all bets are off. There is no logical interposition point at this point.

File: torch/_subclasses/fake_tensor.py (hunk 5 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:667]: It needs to happen early. In particular, it needs to happen before we check for non-fake, as the triggering case was we are in fake tensor mode querying if a non fake tensor is coalesced.

File: torch/_subclasses/meta_utils.py (hunk 6 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:277]: The old code always created a storage, and then assembled it. This new path tries to create the tensors in a more "logical" way, which is using the normal `empty_strided` constructor. However, this constructor goes ahead and creates a storage for you, so you shouldn't try to create one yourself if you're OK with this storage.

File: torch/_subclasses/meta_utils.py (hunk 7 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:329]: I was going to say NotImplemented is annoying to deal with in run_meta_crossref because we run the converter inside of a `tree_map`, but I guess if we return non-Tensor inputs without returning NotImplemented there it is fine, I can still query the miss variable.

File: torch/_subclasses/meta_utils.py (hunk 8 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:322]: not sure how it wouldn't be convertible. Seems like the same thing as asking what if r is not convertible. We... fix it.

File: torch/_subclasses/meta_utils.py (hunk 9 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:371]: yeah, the whole point of the callback argument was to make sure we preserved non-leafness, but now we've gone and wrapped it in another Tensor constructor after we processed everything. But maybe it's ok, cuz parameters are always leafs anyway??

File: test/test_meta.py [hunk 13]
```diff
@@ -66,6 +66,34 @@ def assertSameVersionCounter(self, m1, m2):
...
+        self.assertEqual(m1.is_leaf, m2.is_leaf)
+        self.assertEqual(m1.grad_fn is None, m2.grad_fn is None)
+        self.assertEqual(m1.is_sparse, m2.is_sparse)
+        self.assertEqual(m1.is_inference(), m2.is_inference())
+        self.assertEqual(m1.is_conj(), m2.is_conj())
+        self.assertEqual(m1.is_neg(), m2.is_neg())
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", "The .grad attribute of a Tensor")
+            grad_not_none = m1.grad is not None
+        if grad_not_none:
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_meta.py:82]: yes, good catch

File: test/test_meta.py [hunk 14]
```diff
@@ -66,6 +66,34 @@ def assertSameVersionCounter(self, m1, m2):
...
+            self.assertMetadataMatches(m1.grad, m2.grad)
+        if m1.is_sparse:
+            self.assertEqual(m1.dense_dim(), m2.dense_dim())
+            self.assertEqual(m1.sparse_dim(), m2.sparse_dim())
+            self.assertEqual(m1.is_coalesced(), m2.is_coalesced())
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_meta.py:87]: yes, I suppose we should

File: test/test_meta.py [hunk 15]
```diff
@@ -66,6 +66,34 @@ def assertSameVersionCounter(self, m1, m2):
...
+        else:
+            self.assertEqual(m1.stride(), m2.stride())
+            self.assertEqual(m1.storage_offset(), m2.storage_offset())
+            self.assertEqual(m1._is_view(), m2._is_view())
+            if m1._is_view():
+                self.assertMetadataMatches(m1._base, m2._base)
+        # TODO: test if is resizable (no direct query for this atm)
+        # TODO: audit AutogradMeta to see if it matches
```
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_meta.py:95]: Forward AD agreed.

Hooks, I'm not so sure. Can a hook trigger if it is attached to the INPUT tensor of a grad compute? I guess if it's a pre-hook? But it will trigger anyway when the gradient flows to the original, real tensor. So you might double trigger the hook in that situation.

File: torch/_subclasses/meta_utils.py (hunk 12 above)
Comment (ezyang) [reply torch/_subclasses/meta_utils.py:280]: Hmm, you are right! Do you think it is better if I view and detach in this case, instead of using set_?

**Comment by ezyang** [comment]
Gonna handle comments on follow up

File: test/test_meta.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply test/test_meta.py:87]: Actually, this would fail, because I don't propagate nnz when fakifying. nnz is very data dependent, so it's not clear it's worth propagating.
